elm_deps_check.py

```diff
@@ -4,61 +4,144 @@
 import sys
 import json
 import argparse
-
-
-def have_matching_versions(top_level_file, spec_file, is_exact=False, quiet=True):
-    """ first file should be the top level elm-package exact-dependencies.json
-        second file should be the spec file
+import textwrap
+
+
+class PackageNotFound(object):
+    friendly = textwrap.dedent('''\
+    Package {package_name} was a dependency in the reference package file
+    ({reference_name}) but was not in the candidate package file
+    ({candidate_name}).
+
+    You can probably fix this error by adding {package_name} to the
+    dependencies in {candidate_name}.\
+    ''')
+
+    def __init__(self, package_name, reference_name, candidate_name):
+        self.package_name = package_name
+        self.reference_name = reference_name
+        self.candidate_name = candidate_name
+
+    def __str__(self):
+        return self.friendly.format(
+            package_name=self.package_name,
+            reference_name=self.reference_name,
+            candidate_name=self.candidate_name,
+        )
+
+
+class VersionMismatch(object):
+    friendly = textwrap.dedent('''\
+    Package version mismatch for {package_name}!
+
+    The reference package file ({reference_name}) has version
+    "{reference_version}", but the candidate package file ({candidate_name})
+    has version "{candidate_version}".
+
+    You can probably fix this error by changing the version bounds for
+    {package_name} in {candidate_name} to "{reference_version}".\
+    ''')
+
+    def __init__(
+            self,
+            package_name,
+            reference_name, reference_version,
+            candidate_name, candidate_version):
+        self.package_name = package_name
+        self.reference_name = reference_name
+        self.reference_version = reference_version
+        self.candidate_name = candidate_name
+        self.candidate_version = candidate_version
+
+    def __str__(self):
+        return self.friendly.format(
+            package_name=self.package_name,
+            reference_name=self.reference_name,
+            reference_version=self.reference_version,
+            candidate_name=self.candidate_name,
+            candidate_version=self.candidate_version,
+        )
+
+
+def have_matching_versions(reference_file, candidate_file, is_exact=False, quiet=True):
+    """ first file should be the top level elm-package.json
+        second file should be the spec file elm-package.json
     """
-
-    with open(top_level_file) as f:
-        top_level = json.load(f)
-
-    with open(spec_file) as f:
-        spec = json.load(f)
+    reference = json.load(reference_file)
+    candidate = json.load(candidate_file)
 
     if not is_exact:
-        top_level = top_level['dependencies']
-        spec = spec['dependencies']
+        reference = reference['dependencies']
+        candidate = candidate['dependencies']
```

Comment:
Could you add a comment here to explain how this makes the matching inexact?

Reply:
I'm not sure what you mean here?

Reply:
Some context is lost in the comment view. This block is:

```python
    if not is_exact:
        reference = reference['dependencies']
        candidate = candidate['dependencies']
```

It's not obvious (to me) how this creates inexact behaviour. If I'm just being obtuse then don't worry, but otherwise a comment would help me now, and all of our future selves.

Reply:
`exact` dependencies are those in the elm-stuff folder. In there, they are just a json blob that looks like this:

```
{ elm-lang/html: 1.0.0
, elm-lang/core: 1.0.0
}
```

So they _are_ the dependencies and therefore can be compared directly. This is needed since if you don't compare the exact_dependencies, you can't be certain that the deploys you make are actually the same version. This leads to issues where you have a package in your root project which don't match your test package.

Reply:
Okay, I get it now. Thanks!